<a href="https://colab.research.google.com/github/keishauliiaa-prog/PraktikDasarStatistikPython/blob/main/Latihan_Pengenalan_Data_Wrangling_Python.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 🐍 Analisis Data Shopping Cart Database

### 🚀 Tujuan
*   Mengumpulkan seluruh data yang dibutuhkan.
*   Menilai kualitas dari data yang telah dikumpulkan.
*   Membersihkan data tersebut sehingga siap untuk dianalisis.

### 🔄 Alur Latihan
1. Tahap persiapan.
2. Tahap `gathering` data.
3. Tahap `assessing` data.
4. Tahap `cleanin`g data.

### 📊 Data Sheet : [Kaggle - Shopping Cart](https://www.kaggle.com/datasets/ruchi798/shopping-cart-database)

## 📥 Tahap `Gathering` data

In [ ]:
# Import library
import numpy as np                  # Untuk perhitungan numerik
import pandas as pd                 # Untuk membuat dan mengelola DataFrame
import matplotlib.pyplot as plt     # Untuk visualisasi dasar
import seaborn as sns               # Untuk visualisasi lanjutan

# Koneksi ke Google Drive
from google.colab import drive
drive.mount('/content/drive')       # Memasukkan Google Drive agar file dapat diakses


Mounted at /content/drive


🔹 Keterangan:

* `drive.mount('/content/drive')` → menghubungkan Colab dengan Google Drive agar file CSV bisa diakses.

* Perlu dilakukan sekali per sesi jika ingin membaca file dari Drive.

In [ ]:
# Membuat variabel file path
file_customers = '/content/drive/MyDrive/Dian Folder/Materi - Data Analyst/Camp 02 - Data Analyst/Materi 4 - Pengenalan Data Wrangling/Modul Praktik/customers.csv'  # File customers
file_orders = '/content/drive/MyDrive/Dian Folder/Materi - Data Analyst/Camp 02 - Data Analyst/Materi 4 - Pengenalan Data Wrangling/Modul Praktik/orders.csv'        # File orders
file_products = '/content/drive/MyDrive/Dian Folder/Materi - Data Analyst/Camp 02 - Data Analyst/Materi 4 - Pengenalan Data Wrangling/Modul Praktik/products.csv'    # File products
file_sales = '/content/drive/MyDrive/Dian Folder/Materi - Data Analyst/Camp 02 - Data Analyst/Materi 4 - Pengenalan Data Wrangling/Modul Praktik/sales.csv'          # File sales


🔹 Keterangan Metode:

* Menyimpan path file ke variabel agar mudah digunakan saat membaca file CSV.

* Mempermudah penggantian lokasi file di masa mendatang tanpa mengubah seluruh kode.

In [ ]:
# Membaca data dari file CSV ke dalam DataFrame
customers_df = pd.read_csv(file_customers)   # Membaca data pelanggan
orders_df = pd.read_csv(file_orders)         # Membaca data order
products_df = pd.read_csv(file_products)     # Membaca data produk
sales_df = pd.read_csv(file_sales)           # Membaca data penjualan

🔹 Keterangan:

* `pd.read_csv(file_path)` → membaca file CSV menjadi DataFrame pandas.

* `DataFrame` memudahkan manipulasi data seperti filtering, agregasi, dan analisis statistik.

In [ ]:
# Menampilkan 5 baris pertama setiap DataFrame untuk melihat isi data
print(customers_df.head())    # 5 data pertama pelanggan
print(orders_df.head())       # 5 data pertama order
print(products_df.head())     # 5 data pertama produk
print(sales_df.head())        # 5 data pertama penjualan

   customer_id customer_name  gender  age                    home_address  \
0            1        Nama 1  Female   30  8606 Victoria TerraceSuite 560   
1            2        Nama 2  Female   69      8327 Kirlin SummitApt. 461   
2            3        Nama 3  Female   59       269 Gemma SummitSuite 109   
3            4        Nama 4  Female   67       743 Bailey GroveSuite 141   
4            5        Nama 5  Female   30         48 Hyatt ManorSuite 375   

   zip_code               city                         state    country  
0      5464      Johnstonhaven            Northern Territory  Australia  
1      8223    New Zacharyfort               South Australia  Australia  
2      5661           Aliburgh  Australian Capital Territory  Australia  
3      1729  South Justinhaven                    Queensland  Australia  
4      4032     Griffithsshire                    Queensland  Australia  
   order_id  customer_id  payment order_date delivery_date
0         1           64    30811 

🔹 Keterangan:

* `df.head()` → menampilkan beberapa baris pertama, default 5 baris.

* Berguna untuk cek cepat struktur data, tipe kolom, dan kualitas data awal sebelum analisis.

## 🔍 Tahap `Assessing` Data

### 1️⃣ Menilai Data Customers

In [ ]:
# Menampilkan ringkasan informasi tentang Data Customers
customers_df.info()   # Menampilkan jumlah baris, kolom, tipe data, dan non-null count tiap kolom


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1007 entries, 0 to 1006
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   customer_id    1007 non-null   int64 
 1   customer_name  1007 non-null   object
 2   gender         989 non-null    object
 3   age            1007 non-null   int64 
 4   home_address   1007 non-null   object
 5   zip_code       1007 non-null   int64 
 6   city           1007 non-null   object
 7   state          1007 non-null   object
 8   country        1007 non-null   object
dtypes: int64(3), object(6)
memory usage: 70.9+ KB


🔹 Keterangan:

* `df.info()` → mengecek struktur data dan tipe kolom.

* Bisa langsung melihat kolom mana yang memiliki missing value (n`on-null count` < `total rows`).

🔹 Insight Penting:

* Kolom gender memiliki `18 nilai kosong` → perlu di-handle saat cleaning data.

* Tipe data sudah sesuai (`int64` untuk numerik, `object` untuk teks).

In [ ]:
# Menampilkan jumlah missing value per kolom
customers_df.isna().sum()   # Menghitung jumlah nilai kosong (NaN) tiap kolom

,0
customer_id,0
customer_name,0
gender,18
age,0
home_address,0
zip_code,0
city,0
state,0
country,0


🔹 Keterangan:

* `df.isna().sum()` → menghitung missing value per kolom.

* Berguna untuk mengetahui kolom mana yang perlu diisi/diimputasi sebelum analisis.

🔹 Insight Penting:

* Hanya kolom gender yang memiliki missing value (`18 baris`).

* Kolom lain sudah lengkap → siap untuk analisis/statistik.

In [ ]:
# Mengecek jumlah duplikasi berdasarkan customer_id
duplicate_ids = customers_df.duplicated(subset=['customer_id'])  # Mengecek duplikasi hanya di kolom customer_id
print("Jumlah duplikasi berdasarkan customer_id:", duplicate_ids.sum())  # Menampilkan jumlah duplikasi


Jumlah duplikasi berdasarkan customer_id: 7


🔹 Keterangan:

* `df.duplicated(subset=['colom'])` → menghitung baris duplikat pada kolom tertentu.

* Bisa digunakan untuk membersihkan data ganda agar analisis lebih akurat.

🔹 Insight Penting:

* Terdapat `7 baris duplikat` → sebaiknya dihapus saat data cleaning.

In [ ]:
# Menampilkan ringkasan statistik Data Customers
customers_df.describe()   # Menampilkan statistik numerik: count, mean, std, min, Q1, median, Q3, max

,customer_id,age,zip_code
count,1007.000000,1007.000000,1007.000000
mean,501.726912,50.929494,5012.538232
std,288.673238,30.516299,2885.836112
min,1.000000,20.000000,2.000000
25%,252.500000,34.000000,2403.500000
50%,502.000000,50.000000,5087.000000
75%,751.500000,65.000000,7493.500000
max,1000.000000,700.000000,9998.000000


🔹 Keterangan Metode:

* `df.describe()` → ringkasan statistik untuk kolom numerik.

* Membantu memahami `distribusi`, `outlier`, dan `skala data`.

🔹 Insight Penting:

* `age` → rata-rata 50.9 tahun, rentang `20–700` (nilai 700 tampak anomali → outlier).

* `customer_id` → terlihat terurut dari 1–1000, distribusi normal.

### 2️⃣ Menilai Data Order

In [ ]:
# Menampilkan ringkasan informasi tentang Data Orders
orders_df.info()  # Menampilkan jumlah baris, kolom, tipe data, dan non-null count tiap kolom


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   order_id       1000 non-null   int64 
 1   customer_id    1000 non-null   int64 
 2   payment        1000 non-null   int64 
 3   order_date     1000 non-null   object
 4   delivery_date  1000 non-null   object
dtypes: int64(3), object(2)
memory usage: 39.2+ KB


🔹 Keterangan Metode:

* `df.info()` → mengecek struktur data dan tipe kolom.

* Penting untuk mengetahui apakah ada kolom dengan missing value atau tipe data yang salah.

🔹 Insight Penting:

* Semua kolom `non-null`, berarti tidak ada missing value.

* Kolom `order_date` dan `delivery_date` `bertipe object` → sebaiknya dikonversi ke `datetime` agar mudah dianalisis.

In [ ]:
# Mengecek duplikasi berdasarkan kolom Order_ID
duplicate_id = orders_df.duplicated(subset=['order_id']).sum()  # Menghitung baris dengan Order_ID sama
print("Jumlah duplikasi berdasarkan order_id: ", duplicate_id)

Jumlah duplikasi berdasarkan order_id:  0


🔹 Keterangan Metode:

* `df.duplicated()` → mendeteksi baris yang identik penuh.

* `.sum()` → menghitung jumlah total baris duplikat.

🔹 Insight Penting:

* Hasil 0 → tidak ada baris duplikat → data sudah unik per baris.

In [ ]:
# 🔹 Menampilkan ringkasan statistik Data Orders
orders_df.describe()  # Ringkasan statistik untuk kolom numerik: count, mean, std, min, Q1, median, Q3, max

,order_id,customer_id,payment
count,1000.000000,1000.000000,1000.000000
mean,500.500000,506.640000,33972.936000
std,288.819436,277.115502,14451.609047
min,1.000000,1.000000,10043.000000
25%,250.750000,275.250000,21329.250000
50%,500.500000,515.000000,33697.500000
75%,750.250000,737.250000,46249.000000
max,1000.000000,1000.000000,59910.000000


🔹 Keterangan Metode:

* `df.describe()` → menampilkan ringkasan statistik untuk kolom numerik.

* Membantu memahami distribusi nilai, outlier, dan skala data sebelum analisis lebih lanjut.

🔹 Insight Penting:

* `payment` → rata-rata Rp33.972, rentang Rp10.043–Rp59.910 → ada variasi besar.

* `customer_id` → nilai rata-rata 506, berarti pelanggan cukup merata dalam dataset.

### 3️⃣ Menilai Data Product

In [ ]:
# Menampilkan ringkasan informasi tentang Data Product
products_df.info()  # Menampilkan jumlah baris, kolom, tipe data, dan non-null count tiap kolom


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1266 entries, 0 to 1265
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   product_id    1266 non-null   int64 
 1   product_type  1266 non-null   object
 2   product_name  1266 non-null   object
 3   size          1266 non-null   object
 4   colour        1266 non-null   object
 5   price         1266 non-null   int64 
 6   quantity      1266 non-null   int64 
 7   description   1266 non-null   object
dtypes: int64(3), object(5)
memory usage: 79.3+ KB


🔹 Keterangan Metode:

* `df.info()` → mengecek struktur data, tipe kolom, dan keberadaan missing value.

🔹 Insight Penting:

* Semua kolom tidak memiliki missing value → siap untuk analisis.

* Tipe data sudah sesuai (`int64` untuk numerik, `object` untuk teks).

In [ ]:
# Menampilkan jumlah data duplikat
print("Jumlah duplikasi: ", products_df.duplicated().sum())  # Menghitung baris yang sama persis


Jumlah duplikasi:  6


🔹 Insight Penting:

* Terdapat `6 baris duplikat` → sebaiknya dihapus saat data cleaning agar analisis lebih akurat.

In [ ]:
# Menampilkan ringkasan statistik Data Product
products_df.describe()  # Ringkasan statistik untuk kolom numerik: count, mean, std, min, Q1, median, Q3, max


,product_id,price,quantity
count,1266.000000,1266.000000,1266.000000
mean,627.926540,105.812006,60.138231
std,363.971586,9.715611,11.682791
min,0.000000,90.000000,40.000000
25%,313.250000,95.250000,50.000000
50%,626.500000,109.000000,60.000000
75%,942.750000,114.000000,70.000000
max,1259.000000,119.000000,80.000000


🔹 Insight Penting:

* `price` → rata-rata 105.8, min 90, max 119 → variasi harga produk kecil.

* `quantity` → rata-rata 60 unit, rentang 40–80 → informasi stok tiap produk.

### 4️⃣ Menilai Data Sales

In [ ]:
# Menampilkan ringkasan informasi tentang Data Sales
sales_df.info()  # Menampilkan jumlah baris, kolom, tipe data, dan non-null count tiap kolom

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 6 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   sales_id        5000 non-null   int64  
 1   order_id        5000 non-null   int64  
 2   product_id      5000 non-null   int64  
 3   price_per_unit  5000 non-null   int64  
 4   quantity        5000 non-null   int64  
 5   total_price     4981 non-null   float64
dtypes: float64(1), int64(5)
memory usage: 234.5 KB


🔹 Keterangan Metode:

* `df.info()` → mengecek struktur data dan tipe kolom.

* Bisa langsung mengetahui apakah ada missing value di tiap kolom.

🔹 Insight Penting:

* Semua kolom selain `total_price` lengkap → ada `19 nilai missing` di total_price.

* Tipe data sudah sesuai (`int64` untuk numerik, `float64` untuk total_price).

In [ ]:
# Menampilkan jumlah missing value per kolom
sales_df.isna().sum()  # Menghitung jumlah nilai kosong (NaN) tiap kolom


,0
sales_id,0
order_id,0
product_id,0
price_per_unit,0
quantity,0
total_price,19


🔹 Keterangan:

* `df.isna().sum()` → Menghitung jumlah nilai kosong (NaN)

🔹 Insight Penting:

* `total_price` memiliki `19 missing value` → perlu diisi atau dihapus sebelum analisis.

* Kolom lain tidak memiliki missing value → siap untuk analisis.

In [ ]:
# Menampilkan jumlah data duplikat
print("Jumlah duplikasi: ", sales_df.duplicated().sum())  # Menghitung baris yang sama persis

Jumlah duplikasi:  0


🔹 Insight Penting:

`Hasil 0` → tidak ada baris identik penuh → data unik per baris.

In [ ]:
# Menampilkan ringkasan statistik Data Sales
sales_df.describe()  # Ringkasan statistik untuk kolom numerik: count, mean, std, min, Q1, median, Q3, max

,sales_id,order_id,product_id,price_per_unit,quantity,total_price
count,5000.000000,5000.000000,5000.000000,5000.000000,5000.00000,4981.000000
mean,2499.500000,503.038200,634.053200,103.501600,1.99240,206.307368
std,1443.520003,285.964418,363.255794,9.195004,0.80751,86.352449
min,0.000000,1.000000,1.000000,90.000000,1.00000,90.000000
25%,1249.750000,258.000000,323.000000,95.000000,1.00000,112.000000
50%,2499.500000,504.500000,635.000000,102.000000,2.00000,204.000000
75%,3749.250000,749.000000,951.000000,112.000000,3.00000,285.000000
max,4999.000000,999.000000,1259.000000,119.000000,3.00000,357.000000


🔹 Insight Penting:

* `price_per_unit` → rata-rata 103.5, min 90, max 119 → variasi harga relatif kecil.

* `quantity` → sebagian besar 1–3 unit per transaksi → typical order kecil.

* `total_price` → rata-rata 206.3, min 90, max 357 → ada variasi karena jumlah unit × harga.

* `total_price` memiliki `19 missing` → harus dicek sebelum analisis revenue.

## 🧹 Tahap `Cleaning` Data

### 1️⃣ Membersihkan Data Customers

In [ ]:
# Menghapus Duplicate Data
customers_df.drop_duplicates(subset=['customer_id'], inplace=True)  # Hanya menyisakan satu baris per customer_id
print("Jumlah duplikasi setelah cleaning:", customers_df.duplicated().sum())  # Hasil = 0

Jumlah duplikasi setelah cleaning: 0


🔹 Keterangan:

* `drop_duplicates()` → menghapus baris identik penuh, memastikan setiap baris unik.

* `.duplicated().sum()` → memastikan jumlah duplikasi = 0.

In [ ]:
# Menangani Missing Value
print("\nBaris dengan missing gender:\n", customers_df[customers_df.gender.isna()])  # Menampilkan baris NaN
print("\nDistribusi gender sebelum filling:\n", customers_df.gender.value_counts())  # Melihat distribusi nilai

customers_df.fillna(value="Prefer not to say", inplace=True)  # Mengisi missing gender dengan 'Prefer not to say'
print("\nJumlah missing value setelah filling:\n", customers_df.isna().sum())  # Mengecek kembali, harus 0


Baris dengan missing gender:
      customer_id customer_name gender  age                     home_address  \
38            39       Nama 39    NaN   80  7440 Cameron Estate DrSuite 628   
167          168      Nama 168    NaN   27         2781 Berge MallSuite 452   
322          322      Nama 323    NaN   30        593 Becker CircleApt. 333   
393          393      Nama 394    NaN   34          5158 Levi HillSuite 531   
442          442      Nama 443    NaN   26           5157 Feil RoadApt. 633   
722          720      Nama 723    NaN   40         31 Jordan ParadeApt. 400   
745          743      Nama 746    NaN   57   09 Christopher StreetSuite 967   
773          771      Nama 774    NaN   74     7367 Wright JunctionApt. 773   
798          795      Nama 799    NaN   49          487 Summer MewsApt. 874   
801          798      Nama 802    NaN   56           27 Aiden KnollApt. 875   
825          822      Nama 826    NaN   59        41 Jenkins KnollSuite 438   
859          855     

🔹 Keterangan:

* `fillna(value=...)` → mengganti semua NaN dengan nilai default (`Prefer not to say`).

* `value_counts()` → melihat distribusi kategori sebelum dan sesudah filling.

In [ ]:
# Menangani Inaccurate Value / Outlier
print("\nBaris dengan umur maksimum sebelum cleaning:\n", customers_df[customers_df.age == customers_df.age.max()])  # Cek outlier

customers_df.age.replace(customers_df.age.max(), 70, inplace=True)  # Ganti nilai maksimum (misal 700) menjadi 70
customers_df.age.replace(customers_df.age.max(), 50, inplace=True)  # Ganti nilai maksimum baru (misal 500) menjadi 50



Baris dengan umur maksimum sebelum cleaning:
      customer_id customer_name             gender  age  \
967          961      Nama 968  Prefer not to say  700   

                   home_address  zip_code        city            state  \
967  29 Farrell ParadeSuite 818      6528  New Joseph  South Australia   

       country  
967  Australia  


/tmp/ipython-input-4238700643.py:4: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  customers_df.age.replace(customers_df.age.max(), 70, inplace=True)  # Ganti nilai maksimum (misal 700) menjadi 70
/tmp/ipython-input-4238700643.py:5: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For 

🔹 Keterangan:

* `replace(old, new, inplace=True)` → mengganti nilai tidak realistis di kolom age.

* Membuat data umur lebih masuk akal untuk analisis.

In [ ]:
# Mengecek Statistik Setelah Cleaning
summary_stats = customers_df.describe()  # Ringkasan statistik kolom numerik
print("\nSummary Statistics setelah cleaning:\n", summary_stats)


Summary Statistics setelah cleaning:
        customer_id          age     zip_code
count  1007.000000  1007.000000  1007.000000
mean    501.726912    49.857001  5012.538232
std     288.673238    17.641262  2885.836112
min       1.000000    20.000000     2.000000
25%     252.500000    34.000000  2403.500000
50%     502.000000    50.000000  5087.000000
75%     751.500000    64.500000  7493.500000
max    1000.000000    80.000000  9998.000000


🔹 Keterangan:

* `describe(`) → menampilkan count, mean, std, min, max, median, Q1, Q3.

* Memastikan semua kolom numerik valid dan siap untuk analisis lebih lanjut.

### 2️⃣ Membersihkan Data Order

In [ ]:
# Mengubah tipe data tanggal menjadi datetime
datetime_columns = ["order_date", "delivery_date"]   # Daftar kolom yang akan dikonversi

for column in datetime_columns:                      # Loop untuk setiap kolom tanggal
    orders_df[column] = pd.to_datetime(orders_df[column])  # Konversi tipe object → datetime

In [ ]:
# Menampilkan informasi dataset setelah konversi
orders_df.info()  # Mengecek tipe data kolom, memastikan order_date & delivery_date sudah datetime

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   order_id       1000 non-null   int64         
 1   customer_id    1000 non-null   int64         
 2   payment        1000 non-null   int64         
 3   order_date     1000 non-null   datetime64[ns]
 4   delivery_date  1000 non-null   datetime64[ns]
dtypes: datetime64[ns](2), int64(3)
memory usage: 39.2 KB


🔹 Keterangan:

* `pd.to_datetime()` digunakan untuk mengonversi kolom bertipe object (string) menjadi datetime64

* `info()` digunakan untuk memastikan perubahan tipe data berhasil.

### 3️⃣ Membersihkan Data Product

In [ ]:
# Menghapus data duplikat
products_df.drop_duplicates(subset=['product_id'],inplace=True)  # Menghapus seluruh baris yang identik (duplikat)

# Mengecek kembali jumlah data duplikat setelah pembersihan
print("Jumlah duplikasi: ", products_df.duplicated().sum())  # Memastikan tidak ada duplikat yang tersisa

Jumlah duplikasi:  0


🔹 Keterangan:

* Fungsi `drop_duplicates()` digunakan untuk menghapus baris yang memiliki seluruh nilai kolom identik.

* Parameter `inplace=True` berarti perubahan dilakukan langsung pada DataFrame tanpa membuat salinan baru.

* `duplicated(`) menghasilkan nilai `True` untuk baris yang duplikat, sedangkan `.sum()` menghitung jumlahnya.

### 4️⃣ Membersihkan Data Sales

In [ ]:
# Menampilkan data yang memiliki nilai kosong (NaN) pada kolom 'total_price'
sales_df[sales_df.total_price.isna()]   # Menampilkan baris dengan nilai total_price kosong

,sales_id,order_id,product_id,price_per_unit,quantity,total_price
9,9,2,1196,105,1,NaN
121,121,27,1027,90,3,NaN
278,278,63,360,94,2,NaN
421,421,95,1091,115,1,NaN
489,489,108,1193,105,3,NaN
539,539,117,405,119,2,NaN
636,636,134,653,93,3,NaN
687,687,145,1138,102,1,NaN
854,854,177,64,104,1,NaN
1079,1079,222,908,94,3,NaN


🔹 Keterangan:

* `.isna()` → digunakan untuk mendeteksi missing value.

* Dari hasil tampilan, ditemukan beberapa baris dengan total_price bernilai NaN.

In [ ]:
# Mengisi nilai kosong dengan perhitungan ulang
sales_df["total_price"] = sales_df["price_per_unit"] * sales_df["quantity"]   # Menghitung ulang total_price


🔹 Keterangan:

* Operasi antar kolom pada `DataFrame` (kolom dapat langsung dikalikan)
* Karena `'total_price'` = `'price_per_unit'` × `'quantity'`, maka perhitungan ini memperbaiki data hilang.

In [ ]:
# Memastikan tidak ada lagi missing value
sales_df.isna().sum()   # Mengecek jumlah nilai NaN per kolom

,0
sales_id,0
order_id,0
product_id,0
price_per_unit,0
quantity,0
total_price,0


🔹 Keterangan:

* `.isna().sum()` → menghitung jumlah nilai kosong di setiap kolom.
* Semua kolom sekarang tidak memiliki nilai NaN, artinya data sudah bersih.

# 💾 Menyimpan data yang sudah dibersihkan ke file CSV

In [ ]:
# Membuat folder output (opsional, untuk kerapihan)
import os
output_path = "/content/drive/MyDrive/Dian Folder/Materi - Data Analyst/Camp 02 - Data Analyst/Materi 4 - Pengenalan Data Wrangling/Modul Praktik/Data_Cleaned"
os.makedirs(output_path, exist_ok=True)

# Simpan setiap dataset yang sudah dibersihkan
customers_df.to_csv(f"{output_path}/customers_cleaned.csv", index=False)
orders_df.to_csv(f"{output_path}/orders_cleaned.csv", index=False)
products_df.to_csv(f"{output_path}/products_cleaned.csv", index=False)
sales_df.to_csv(f"{output_path}/sales_cleaned.csv", index=False)

print("✅ Semua data berhasil disimpan ke folder:")
print(output_path)

✅ Semua data berhasil disimpan ke folder:
/content/drive/MyDrive/Dian Folder/Materi - Data Analyst/Camp 02 - Data Analyst/Materi 4 - Pengenalan Data Wrangling/Modul Praktik/Data_Cleaned


🔹 Penjelasan Kode

* o`s.makedirs(path, exist_ok=True)` → membuat folder baru; `exist_ok=True` mencegah error jika folder sudah ada.

* `to_csv(..., index=False)` → menyimpan DataFrame ke file CSV tanpa kolom indeks agar lebih rapi.

* Folder output dipisahkan agar file bersih mudah ditemukan dan dikelola.

* `print(output_path)` → menampilkan path folder hasil simpan, memudahkan navigasi di Google Drive/Colab.

# 🔀 Menggabungkan Data: Merge / Join

In [ ]:
# Inner Join (hanya mengambil data yang cocok di kedua tabel)
inner_join_df = pd.merge(customers_df, orders_df, on="customer_id", how="inner")  # Hanya customer yang punya order
print("Inner Join (customer yang punya order):")
print(inner_join_df.head(), "\n")

# Left Join (ambil semua data dari kiri + data cocok dari kanan)
left_join_df = pd.merge(customers_df, orders_df, on="customer_id", how="left")  # Semua customer, order bisa NaN
print("Left Join (semua customer, order mungkin kosong):")
print(left_join_df.head(), "\n")

# Right Join (ambil semua data dari kanan + data cocok dari kiri)
right_join_df = pd.merge(customers_df, orders_df, on="customer_id", how="right")  # Semua order, customer mungkin kosong
print("Right Join (semua order, customer mungkin kosong):")
print(right_join_df.head(), "\n")

# Outer Join (ambil semua data dari kedua tabel, gabungkan jika cocok)
outer_join_df = pd.merge(customers_df, orders_df, on="customer_id", how="outer")  # Semua customer dan order
print("Outer Join (semua customer dan order):")
print(outer_join_df.head(), "\n")


Inner Join (customer yang punya order):
   customer_id customer_name  gender  age                    home_address  \
0            1        Nama 1  Female   30  8606 Victoria TerraceSuite 560   
1            1        Nama 1  Female   30  8606 Victoria TerraceSuite 560   
2            1        Nama 1  Female   30  8606 Victoria TerraceSuite 560   
3            7        Nama 7  Female   76     92 Hills Station StApt. 683   
4           10       Nama 10  Female   70       615 Hayley KnollSuite 454   

   zip_code           city               state    country  order_id  payment  \
0      5464  Johnstonhaven  Northern Territory  Australia       455    24550   
1      5464  Johnstonhaven  Northern Territory  Australia       670    10246   
2      5464  Johnstonhaven  Northern Territory  Australia       729    35593   
3       793      Masonfurt          Queensland  Australia       465    48935   
4      2118      Joelburgh   Western Australia  Australia       595    45626   

  order_date del

🔹 Penjelasan Metode Join

1. `on="customer_id"` → kolom yang digunakan untuk menggabungkan kedua DataFrame.

2. how parameter:

* `"inner"` → hanya baris yang cocok di kedua DataFrame.

* `"left"` → semua baris dari DataFrame kiri (customers_df) + data cocok dari kanan.

* `"right"` → semua baris dari DataFrame kanan (orders_df) + data cocok dari kiri.

* `"outer"` → semua baris dari kedua DataFrame, gabungkan jika ada kecocokan.